community/values/src/main/java/org/neo4j/values/storable/StringValue.java

```diff
@@ -55,7 +52,56 @@ public boolean equals( String x )
     @Override
     public int computeHash()
     {
-        return value().hashCode();
+        //NOTE that we are basing the hash code on code points instead of char[] values.
+        String value = value();
+        if ( value.isEmpty() )
+        {
+            return 0;
+        }
+        int h = 1;
+        final int length = value.length();
+        for ( int offset = 0; offset < length; )
+        {
+            final int codePoint = value.codePointAt( offset );
+            h = 31 * h + codePoint;
+            offset += Character.charCount( codePoint );
+        }
+        return h;
+    }
+
+    @Override
+    public TextValue substring( int start, int end )
```

Comment:
I'm wondering if we should push the implementation of these methods down to the leaf types, so instead of `StringValue` providing a default implementation that `UTF8StringValue` overrides, we would have `UTF8StringValue` and `StringWrappingStringValue` each simply provide their own implementation.